p2p/chain.py

```diff
@@ -164,142 +149,134 @@ def register_peer(self, peer: BasePeer) -> None:
-                if len(batch) >= batch_size:
-                    await request_func(batch[:batch_size])
-                    batch = batch[batch_size:]
-
-            await self._retry_timedout(request_func, pending, batch_size, part_name)
-
-    async def _retry_timedout(self,
-                              request_func: Callable[[List[BlockHeader]], Awaitable[None]],
-                              pending: Dict[Any, Tuple[BlockHeader, float]],
-                              batch_size: int,
-                              part_name: str) -> None:
-        now = time.time()
-        timed_out = [
-            header
-            for header, req_time
-            in pending.values()
-            if now - req_time > self._reply_timeout
-        ]
-        while timed_out:
-            self.logger.warn(
-                "Re-requesting %d timed out %s out of %d pending ones",
-                len(timed_out), part_name, len(pending))
-            await request_func(timed_out[:batch_size])
-            timed_out = timed_out[batch_size:]
-
-    async def body_downloader(self) -> None:
-        await self._downloader(
-            self._body_requests,
-            self._skip_empty_bodies,
-            self.request_bodies,
-            self._pending_bodies,
-            eth.MAX_BODIES_FETCH,
-            'bodies')
-
-    async def receipt_downloader(self) -> None:
-        await self._downloader(
-            self._receipt_requests,
-            self._skip_empty_and_duplicated_receipts,
-            self.request_receipts,
-            self._pending_receipts,
-            eth.MAX_RECEIPTS_FETCH,
-            'receipts')
-
-    @to_list
-    def _skip_empty_bodies(self, headers: List[BlockHeader]) -> Generator[BlockHeader, None, None]:
-        for header in headers:
-            if (header.transaction_root != self.chaindb.empty_root_hash or
-                    header.uncles_hash != EMPTY_UNCLE_HASH):
-                yield header
-
-    async def request_bodies(self, headers: List[BlockHeader]) -> None:
-        peer = await self.get_idle_peer()
-        peer.sub_proto.send_get_block_bodies([header.hash for header in headers])
-        self._peers_with_pending_requests.add(peer)
+                pending_replies = request_func(missing)
+                continue
+
+            pending_replies -= 1
+            unexpected = set(downloaded).difference(
+                [key_func(header) for header in missing])
+            for item in unexpected:
+                self.logger.warn("Got unexpected %s: %s", part_name, unexpected)
+            missing = [
+                header for header in missing
+                if key_func(header) not in downloaded]
+
+    def _request_block_parts(
+            self,
+            headers: List[BlockHeader],
+            request_func: Callable[[ETHPeer, List[BlockHeader]], None]) -> int:
+        length = math.ceil(len(headers) / len(self.peer_pool.peers))
+        batches = list(partition_all(length, headers))
```

Comment:
I assume the `list` casting here has to do with the following `zip` not supporting whatever `partition_all` natively returns?

Reply:
Yep, otherwise we can't get its length